File: api/schemas/student.py
"""
api/schemas/student.py
======================
Pydantic schemas para los endpoints del estudiante.
"""

from pydantic import BaseModel, Field


# ── Preguntas ─────────────────────────────────────────────────────────────────


class NextQuestionRequest(BaseModel):
    course_id: str = Field(..., description="Curso activo del estudiante")
    topic: str | None = Field(default=None, description="Tópico específico (opcional)")
    block: str | None = Field(default=None, description="Bloque temático (DIAN/SENA)")
    session_correct_ids: list[str] = Field(
        default_factory=list,
        description="IDs de ítems respondidos correctamente en esta sesión",
    )
    session_wrong_timestamps: dict[str, int] = Field(
        default_factory=dict,
        description="item_id → pregunta_num en que fue fallada en sesión",
    )
    session_questions_count: int = Field(default=0, ge=0)


class ItemResponse(BaseModel):
    id: str
    content: str
    difficulty: float
    topic: str
    options: list[str]
    image_url: str | None = None
    tags: list[str] = []


class NextQuestionResponse(BaseModel):
    item: ItemResponse | None
    status: str  # "ok", "empty", "course_empty"


# ── Respuestas ────────────────────────────────────────────────────────────────


class AnswerRequest(BaseModel):
    item_id: str = Field(..., description="ID del ítem respondido")
    item_data: dict | None = Field(
        default=None,
        description="Compatibilidad con clientes anteriores; el servidor ignora estos datos.",
        deprecated=True,
    )
    selected_option: str = Field(..., description="Opción elegida por el estudiante")
    reasoning: str | None = Field(
        default="", description="Razonamiento del estudiante (para KatIA)"
    )
    time_taken: float | None = Field(default=None, ge=0, description="Segundos en responder")
    elo_topic: str | None = Field(
        default=None,
        description="Tópico del ítem o su course_id; validado contra la base de datos.",
    )


class AnswerResponse(BaseModel):
    is_correct: bool
    elo_before: float
    elo_after: float
    rd_after: float
    delta_elo: float
    cog_data: dict


# ── Stats y ELO ──────────────────────────────────────────────────────────────


class TopicELO(BaseModel):
    topic: str
    rating: float
    rd: float


class StudentStatsResponse(BaseModel):
    user_id: int
    global_elo: float
    topic_elos: list[TopicELO]
    total_attempts: int
    study_streak: int
    rank_label: str | None = None


# ── Cursos y matrículas ───────────────────────────────────────────────────────


class CourseResponse(BaseModel):
    id: str
    name: str
    block: str
    enrolled: bool
    group_id: int | None = None
    diagnostic_done: bool = False


class EnrollRequest(BaseModel):
    course_id: str
    group_id: int | None = None


class EnrollByCodeRequest(BaseModel):
    invite_code: str


# ── Procedimientos ────────────────────────────────────────────────────────────


class ProcedureSubmitResponse(BaseModel):
    submission_id: int
    ai_score: float | None
    ai_feedback: str | None
    status: str


# ── Modo Examen ───────────────────────────────────────────────────────────────


class ExamStartRequest(BaseModel):
    course_id: str = Field(..., description="Curso para el examen")
    n_questions: int = Field(default=10, ge=1, le=30, description="Número de preguntas (máx 30)")
    time_limit_minutes: int = Field(
        default=20, ge=5, le=180, description="Minutos disponibles para el examen"
    )
    template_id: int | None = Field(
        default=None,
        description=(
            "ID de plantilla de examen creada por el docente. Si se proporciona, "
            "se usa esa lista de items y se ignoran n_questions y time_limit_minutes. "
            "Si es None, se genera un examen estándar con curva de dificultad."
        ),
    )


class ExamTemplateSummary(BaseModel):
    """Resumen visible al estudiante: solo título + nº de preguntas + tiempo."""

    id: int
    title: str
    course_id: str
    n_questions: int
    time_limit_min: int
    created_at: str
    window_ends_at: str | None = None  # si la asignación tiene ends_at, lo expone


class PendingExam(BaseModel):
    """Resumen para el badge de notificación en el sidebar."""

    template_id: int
    title: str
    course_id: str
    course_name: str
    time_limit_min: int


class ExamStartResponse(BaseModel):
    session_id: str
    items: list[ItemResponse]
    n_questions: int
    time_limit_seconds: int
    course_id: str


class ExamAnswerItem(BaseModel):
    item_id: str
    selected_option: str
    time_taken: float | None = None


class ExamSubmitRequest(BaseModel):
    session_id: str = Field(..., min_length=1, max_length=64)
    answers: list[ExamAnswerItem] = Field(..., description="Respuestas del estudiante")
    total_time_taken: float | None = Field(None, description="Tiempo total en segundos")
    course_id: str = Field(default="", description="ID del curso examinado")
    course_name: str = Field(default="", description="Nombre del curso examinado")
    template_id: int | None = Field(
        default=None, description="Plantilla del docente (para análisis de resultados)"
    )


class ExamSubmitResponse(BaseModel):
    results: list[dict]
    correct_count: int
    total_questions: int
    score_pct: float
    global_elo_after: float


# ── Examen diagnóstico (inicio de materia) ────────────────────────────────────


class DiagnosticQuestion(BaseModel):
    id: str
    content: str
    topic: str | None = None
    difficulty: int = 1000
    options: list[str] = Field(default_factory=list)


class DiagnosticStatusResponse(BaseModel):
    completed: bool
    course_id: str
    course_name: str = ""
    questions: list[DiagnosticQuestion] | None = None
    result: dict | None = None


class DiagnosticAnswerItem(BaseModel):
    item_id: str
    selected_option: str = ""  # "" = no contestada / "no lo sé"


class DiagnosticSubmitRequest(BaseModel):
    answers: list[DiagnosticAnswerItem] = Field(default_factory=list)
    course_name: str = ""


class DiagnosticResultResponse(BaseModel):
    initial_elo: float
    score_pct: float
    league: dict
    themes: list[dict]
    correct_total: int
    answered: int
    completed: bool = True


# ── Mapa de contenido (nodos = tópicos del curso) ─────────────────────────────


class MapNode(BaseModel):
    topic: str
    label: str = ""
    label_key: str | None = None
    node_id: str | None = None
    node_type: str = "practice"
    elo: float
    rd: float
    item_count: int
    state: str  # completed | current | available | blocked


class CourseMapResponse(BaseModel):
    course_id: str
    course_name: str = ""
    diagnostic_done: bool
    nodes: list[MapNode]


class LessonEventRequest(BaseModel):
    event: str = Field(
        ...,
        description=(
            "node_viewed | objectives_viewed | math_convention_viewed | node_completed"
        ),
    )


class LessonInteractionRequest(BaseModel):
    interaction_id: str = Field(..., min_length=1, max_length=80)
    selected_option: str = Field(..., min_length=1, max_length=160)


class LessonInteractionResponse(BaseModel):
    interaction_id: str
    selected_option: str
    is_expected: bool | None = None
    misconception_tag: str | None = None
    feedback_key: str


class LessonDetailResponse(BaseModel):
    node_id: str
    node_type: str
    course_id: str
    i18n_prefix: str
    next_node_id: str | None = None
    state: str
    presentation: str
    explored_complex_branch: bool = False
    affects_elo: bool
    is_safe_zone: bool
    optional_branch: bool
    objectives: list[str]
    optional_objectives: list[str]
    interactions: list[dict] = Field(default_factory=list)
    staircase: dict | None = None
    content: dict | None = None
    unlock_after: str | None = None
    progress: dict


# ── Socrático ─────────────────────────────────────────────────────────────────


class SocraticRequest(BaseModel):
    item_id: str
    item_content: str
    student_message: str = Field(..., min_length=1, max_length=1000)
    course_id: str | None = None
    api_key: str = Field(
        default="",
        description="API key del proveedor de IA (opcional — usa la del sistema si vacía)",
    )
    provider: str = Field(default="groq", description="Proveedor: groq, anthropic, openai, etc.")
    lang: str = Field(default="es", description="Idioma de la respuesta de KatIA: 'es' o 'en'.")

File: frontend/src/api/student.ts
/**
 * api/student.ts
 * ==============
 * Clientes tipados para los endpoints del estudiante.
 */

import { api } from "./client";

export interface Item {
  id: string;
  content: string;
  difficulty: number;
  topic: string;
  options: string[];
  image_url?: string;
  tags?: string[];
}

export interface NextQuestionResponse {
  item: Item | null;
  status: "ok" | "empty" | "course_empty";
}

export interface AnswerResponse {
  is_correct: boolean;
  elo_before: number;
  elo_after: number;
  rd_after: number;
  delta_elo: number;
  cog_data: Record<string, unknown>;
}

export interface TopicELO {
  topic: string;
  rating: number;
  rd: number;
}

export interface StudentStats {
  user_id: number;
  global_elo: number;
  topic_elos: TopicELO[];
  total_attempts: number;
  study_streak: number;
  rank_label: string | null;
}

export interface Course {
  id: string;
  name: string;
  block: string;
  enrolled: boolean;
  group_id?: number;
  diagnostic_done?: boolean;
}

// ── Examen diagnóstico (inicio de materia) ──────────────────────────────────
export interface DiagnosticQuestion {
  id: string;
  content: string;
  topic: string | null;
  difficulty: number;
  options: string[];
}
export interface DiagnosticTheme {
  topic: string;
  correct: number;
  total: number;
  ratio: number;
  status: "strong" | "mid" | "gap";
  elo: number;
}
export interface DiagnosticLeague {
  name: string;
  min: number;
  color: string;
  rank: string;
}
export interface DiagnosticResult {
  initial_elo: number;
  score_pct: number;
  league: DiagnosticLeague;
  themes: DiagnosticTheme[];
  correct_total: number;
  answered: number;
  completed?: boolean;
}
export interface DiagnosticStatus {
  completed: boolean;
  course_id: string;
  course_name?: string;
  questions?: DiagnosticQuestion[];
  result?: DiagnosticResult | null;
}
export interface DiagnosticAnswer {
  item_id: string;
  selected_option: string;
}

// ── Mapa de contenido ───────────────────────────────────────────────────────
export interface MapNode {
  topic: string;
  label: string;
  label_key?: string | null;
  node_id?: string | null;
  node_type?: string;
  elo: number;
  rd: number;
  item_count: number;
  state: "completed" | "current" | "available" | "blocked";
}
export interface CourseMap {
  course_id: string;
  course_name: string;
  diagnostic_done: boolean;
  nodes: MapNode[];
}

export type LessonEvent =
  | "node_viewed"
  | "objectives_viewed"
  | "math_convention_viewed"
  | "node_completed";

// N4 (Divisibilidad): un mismo nodo mezcla numeric/single_select/multi_select,
// a diferencia de N2 (solo numeric) o N3 (numeric/text_exact).
export interface N4PracticeItem {
  id: string;
  kind: "numeric" | "single_select" | "multi_select";
  prompt: string;
  expr?: string;
  answer?: string;
  options?: Array<{ id: string; text: string }>;
  valid_options?: string[];
  story?: string;
  image_slot?: boolean;
  image?: string;
  support_objects?: string[];
}

export interface LessonDetail {
  node_id: string;
  node_type: string;
  course_id: string;
  i18n_prefix: string;
  next_node_id: string | null;
  state: string;
  presentation: "basico" | "intermedio" | "avanzado";
  explored_complex_branch: boolean;
  affects_elo: boolean;
  is_safe_zone: boolean;
  optional_branch: boolean;
  objectives: string[];
  optional_objectives: string[];
  unlock_after: string | null;
  interactions: Array<{
    interaction_id: string;
    type: "single_select" | "numeric_input" | "multi_select" | "number_line";
    prompt_key: string;
    option_keys: string[];
    can_retry?: boolean;
  }>;
  content?: {
    kind:
      | "operation_city_hub"
      | "operation_building"
      | "property_laboratory_hub"
      | "property_machine"
      | "level_hub_port"
      | "level_hub_cards"
      | "divisibility_concept";
    level?: string;
    title: string;
    welcome_text?: string;
    scene_text?: string;
    drag_rule?: string;
    advance_text?: string;
    character?: string;
    intro?: string;
    definition?: string;
    formal_expression?: string;
    instruction?: string;
    closing?: string;
    definition_katex?: string;
    machine_id?: string;
    property?: string;
    story_contract?: {
      type: "narrative_with_integrated_definition" | "unified_set_extension" | "guided_discovery_formalization";
      practice_position: string;
      is_integrated: boolean;
    };
    katia?: { eyebrow: string; title: string; body: string; question: string; imageSrc?: string };
    discovery?: { eyebrow: string; title: string; body: string };
    definition_title?: string;
    closure?: {
      eyebrow?: string;
      title: string;
      intro: string;
      rows: Array<{ symbol: string; name: string; closed: "yes" | "no" | "partial"; latex?: string; note: string }>;
    };
    narrative_sections?: Array<{
      type: "prose" | "guided_discovery" | string;
      title: string;
      body: string;
    }>;
    opening_hook?: {
      demo: string;
      katia_message: string;
    };
    validation_status?: string;
    operation?: string;
    gating?: { rule: string; cards_required?: number; machines_required?: number };
    // Textos del hub, parametrizados: sin estos el renderer cae al vocabulario
    // del puerto de la Polis, que es de donde salio.
    image?: string;
    cards_hint?: string;
    cards_aria?: string;
    card_closed_hint?: string;
    card_cta?: string;
    gating_label?: string;
    finish_label?: string;
    buildings?: Array<{
      id: string;
      operation: string;
      symbol: string;
      node_id: string;
      card: string;
    }>;
    machines?: Array<{
      id: string;
      property: string;
      station?: string;
      symbol: string;
      node_id: string;
      demo: string;
      katia_message: string;
      state?: "completed" | "current" | "available" | "blocked";
    }>;
    cards?: Array<{
      id: string;
      concept: string;
      symbol: string;
      node_id: string;
      destination: string;
      teaser: string;
      state?: "completed" | "current" | "available" | "blocked";
    }>;
    icebreaker?: {
      title: string;
      intro: string;
      items: N4PracticeItem[];
    };
    worked_examples?: Array<{
      statement: string;
      solution: string;
      steps?: string[];
      eyebrow?: string;
      title?: string;
      latex?: string;
      image_slot?: boolean;
      image?: string;
      trap?: boolean;
    }>;
    situations?: Array<{ id: string; prompt: string; expr: string; answer: string; set_label?: string }>;
    operation_groups?: Array<{
      id: string;
      katia_after: string;
      ops: Array<{ id: string; expr: string; answer: string }>;
    }>;
    practice?: N4PracticeItem[];
    feedback?: Record<string, string>;
    formalization?:
      | string[]
      | {
          title: string;
          intro: string;
          items: Array<{ label: string; rule: string; latex?: string }>;
        };
  } | null;
  staircase: {
    core_steps: string[];
    optional_detour: string;
  } | null;
  progress: {
    state: string;
    objectives_viewed: boolean;
    math_convention_viewed: boolean;
    viewed_at: string | null;
    completed_at: string | null;
    responses: Record<
      string,
      {
        interaction_id: string;
        selected_option: string;
        is_expected: boolean | null;
        misconception_tag: string | null;
      }
    >;
  };
}

export interface PrealgebraSummary {
  course_id: string;
  completed_nodes: number;
  total_nodes: number;
  overall_status: "strong" | "review" | "attention";
  review: Array<{ node_id: string; label_key: string }>;
}

export interface LessonInteractionResult {
  interaction_id: string;
  selected_option: string;
  is_expected: boolean | null;
  misconception_tag: string | null;
  feedback_key: string;
}

export interface ProcedureSubmissionRow {
  submission_id: number;
  item_id: string;
  item_content: string | null;
  status: string;
  ai_proposed_score: number | null;
  teacher_score: number | null;
  final_score: number | null;
  teacher_feedback: string | null;
  elo_delta: number | null;
  submitted_at: string | null;
  reviewed_at: string | null;
}

export interface ProcedureStep {
  numero?: number;
  contenido?: string;
  evaluacion?: string;
  comentario?: string;
}

export interface ProcedureReview {
  corresponde_a_pregunta?: boolean;
  transcripcion?: string;
  pasos?: ProcedureStep[];
  errores_detectados?: string[];
  saltos_logicos?: string[];
  resultado_correcto?: boolean;
  evaluacion_global?: string;
  score_procedimiento?: number;
}

export interface ExamSession {
  id: number;
  course_id: string;
  course_name: string;
  n_questions: number;
  correct_count: number;
  score_pct: number;
  global_elo_after: number;
  created_at: string;
}

export interface ExamTemplateSummary {
  id: number;
  title: string;
  course_id: string;
  n_questions: number;
  time_limit_min: number;
  created_at: string;
  window_ends_at?: string | null;
}

export interface PendingExam {
  template_id: number;
  title: string;
  course_id: string;
  course_name: string;
  time_limit_min: number;
}

export const studentApi = {
  nextQuestion: (body: {
    course_id: string;
    topic?: string;
    block?: string;
    session_correct_ids?: string[];
    session_wrong_timestamps?: Record<string, number>;
    session_questions_count?: number;
  }) => api.post<NextQuestionResponse>("/api/student/next-question", body),

  courseBlocks: (course_id: string) =>
    api.get<{ block: string; item_count: number }[]>(`/api/student/blocks/${course_id}`),

  pvpHistory: () =>
    api.get<
      Array<{
        won: boolean;
        draw: boolean;
        my_score: number;
        opp_score: number;
        elo_delta: number;
        opponent: string;
        finished_at: string;
      }>
    >("/api/student/pvp/history"),

  answer: (body: {
    item_id: string;
    selected_option: string;
    reasoning?: string;
    time_taken?: number;
    elo_topic?: string;
  }, idempotencyKey: string) => api.postWithHeaders<AnswerResponse>(
    "/api/student/answer", body, { "Idempotency-Key": idempotencyKey },
  ),

  stats: () => api.get<StudentStats>("/api/student/stats"),

  courses: () => api.get<Course[]>("/api/student/courses"),

  enroll: (course_id: string, group_id?: number) =>
    api.post<{ message: string }>("/api/student/enroll", { course_id, group_id }),

  enrollByCode: (invite_code: string) =>
    api.post<{ message: string; course_id: string }>("/api/student/enroll-by-code", {
      invite_code,
    }),

  unenroll: (course_id: string) => api.delete<void>(`/api/student/enroll/${course_id}`),

  diagnostic: (course_id: string, redo = false) =>
    api.get<DiagnosticStatus>(
      `/api/student/diagnostic/${encodeURIComponent(course_id)}${redo ? "?redo=true" : ""}`
    ),

  submitDiagnostic: (course_id: string, answers: DiagnosticAnswer[], course_name?: string) =>
    api.post<DiagnosticResult>(`/api/student/diagnostic/${encodeURIComponent(course_id)}/submit`, {
      answers,
      course_name: course_name ?? "",
    }),

  courseMap: (course_id: string) =>
    api.get<CourseMap>(`/api/student/map/${encodeURIComponent(course_id)}`),

  lesson: (course_id: string, node_id: string) =>
    api.get<LessonDetail>(
      `/api/student/lessons/${encodeURIComponent(course_id)}/${encodeURIComponent(node_id)}`,
    ),

  lessonEvent: (course_id: string, node_id: string, event: LessonEvent) =>
    api.post<LessonDetail>(
      `/api/student/lessons/${encodeURIComponent(course_id)}/${encodeURIComponent(node_id)}/events`,
      { event },
    ),

  lessonInteraction: (
    course_id: string,
    node_id: string,
    interaction_id: string,
    selected_option: string,
  ) =>
    api.post<LessonInteractionResult>(
      `/api/student/lessons/${encodeURIComponent(course_id)}/${encodeURIComponent(node_id)}/interactions`,
      { interaction_id, selected_option },
    ),

  prealgebraSummary: (course_id: string) =>
    api.get<PrealgebraSummary>(
      `/api/student/prealgebra-summary/${encodeURIComponent(course_id)}`,
    ),

  myProcedures: () =>
    api.get<{ submissions: ProcedureSubmissionRow[] }>("/api/student/procedures"),

  reportProblem: (description: string) =>
    api.post<{ message: string }>("/api/student/problems", { description }),

  aiStatus: () =>
    api.get<{ available: boolean; provider: string | null }>("/api/student/ai-status"),

  analyzeProcedure: async (params: {
    item_id: string;
    item_content: string;
    api_key?: string;
    file: File;
  }): Promise<{ review: ProcedureReview; provider: string; analysisToken: string }> => {
    const fd = new FormData();
    fd.append("item_id", params.item_id);
    fd.append("item_content", params.item_content);
    if (params.api_key) fd.append("api_key", params.api_key);
    fd.append("file", params.file);
    const res = await api.postForm<{
      item_id: string;
      provider: string;
      review: ProcedureReview;
      analysis_token: string;
    }>("/api/student/procedure/analyze", fd);
    return { review: res.review, provider: res.provider, analysisToken: res.analysis_token };
  },

  history: () => api.get<{ attempts: unknown[] }>("/api/student/history"),

  examHistory: () => api.get<ExamSession[]>("/api/student/exam/history"),

  examTemplates: (course_id: string) =>
    api.get<ExamTemplateSummary[]>(
      `/api/student/exam/templates?course_id=${encodeURIComponent(course_id)}`,
    ),

  examPending: () => api.get<PendingExam[]>("/api/student/exam/pending"),

  activity: (days = 70) =>
    api.get<{ activity: Record<string, number> }>(`/api/student/activity?days=${days}`),

  streakByCourse: (course_id: string) =>
    api.get<{ course_id: string; streak: number }>(`/api/student/streak/${course_id}`),

  groupRanking: (course_id?: string) => {
    const q = course_id ? `?course_id=${course_id}` : "";
    return api.get<{ ranking: unknown[]; my_rank: number | null }>(`/api/student/group-ranking${q}`);
  },
};
